connectors/src/lib/data_sources.ts

```diff
@@ -229,43 +232,112 @@ async function _updateDocumentParentsField({
   }
 }
 
-const MAX_SECTION_PREFIX_LENGTH = 192;
+// allows for 4 full prefixes before hitting half of the max chunk size (approx.
+// 256 chars for 512 token chunks)
+const MAX_PREFIX_TOKENS = EMBEDDING_CONFIG.max_chunk_size / 8;
+// Limit on chars to avoid tokenizing too much text uselessly on documents with
+// large prefixes. The final truncating will rely on MAX_PREFIX_TOKENS so this
+// limit can be large and should be large to avoid underusing prexfixes
+const MAX_PREFIX_CHARS = MAX_PREFIX_TOKENS * 8;
 
 // The role of this function is to create a prefix from an arbitrary long string. The prefix
 // provided will not be augmented with `\n`, so it should include appropriate carriage return. If
-// the prefix is too long (>256 chars), it will be truncated. The remained will be returned as
+// the prefix is too long (> MAX_PREFIX_TOKENS), it will be truncated. The remained will be returned as
 // content of the resulting section.
-export function renderPrefixSection(
+export async function renderPrefixSection(
+  dataSourceConfig: DataSourceConfig,
   prefix: string | null
-): CoreAPIDataSourceDocumentSection {
+): Promise<CoreAPIDataSourceDocumentSection> {
   if (!prefix || !prefix.trim()) {
     return {
       prefix: null,
       content: null,
       sections: [],
     };
   }
-  if (prefix.length > MAX_SECTION_PREFIX_LENGTH) {
+
+  // if too many chars in the prefix, we have to split AND check for tokens
+  if (prefix.length > MAX_PREFIX_CHARS) {
```

Comment:
As per IRL I think this can be made clearer to follow (right now an else is missing as well but irrelevant if we move with the other solution):

```
let targetPrefix = prefix.substring(0, MAX_PREFIX_CHARS);
let targetContent = (prefix.length > MAX_PREFIX_CHARS) ? prefix.substring(MAX_PREFIX_CHARS) : ""

const tokens = (
      await tokenize(targetPrefix, dataSourceConfig)
);

If (tokens.length > MAX_PREFIX_TOKENS) {
  targetPrefix = tokens
      .slice(0, MAX_PREFIX_TOKENS)
      .map((t) => t[1]).join("");
   targetContent = tokens.slice(MAX_PREFIX_TOKENS).map((t) => t[1]).join("") + targetContent
}

return {
      prefix: targetContent ? targetPrefix + "..." : targetPrefix
      content: targetContent? "..." + targetContent : null
      sections: [],
};
```